### PR TITLE
[Draft] Bar graph using webview and react

### PR DESCRIPTION
**Pull request by parkjisu6239**
This draft is intended to share the current development progress share and the vscode extension file structure.

# Preview
one-preview

# Before you go in...
**We use React.**
We thought that in order to make graphs such as chrome tracing, we need to separate the components and manage them efficiently. We also thought that we needed a more organized and structured library to create node graphs that we were going to add later, or for more functionality to be developed in the future.

So in our code, we are supposed to display the screen that was developed as a react in the webview. And in this process, **_there is a huge difference from the basic folder structure._**

So for Draft to avoid conflict, I temporarily created a folder called BarChart under the src folder and put the changes in it.

# File Structure
The following is a description of the files created under src/BarChart

- `src` and `public` are written as react. When you build content created with react, the build file is created. 
- Settings related to this are in package.json.
- extension.ts is located in the ext-src folder. Here are the settings for extension's command and the connection to the web view.

# Run
(❗ **important** ❗) If you do not have a package manager including node.js, please follow the steps below.
1. install `nodejs`
2. `$ npm install yarn`
3. `$ yarn`

If the above work has been completed or is already installed, you can do the following work.
1. Move the directory to `cd src/BarChart`.
2. Enter `yarn` to create `node_modules`.
3. Enter `yarn watch-build` to create a `build` file.
4. When the phrase `[nodeemon] clean exit - waiting for changes before restart` appears, press `ctrl + c` to exit the build.
5. Press `F5` to run extension.
6. When a warning window appears that starts with `The task~`, press 'Debug anyway' to move on.
7. When a new window in vscode opens, press `ctrl + shift + P` to enter `ONE: one-vscode-extension` command.
img1
img2
img3

# Developed Features
- Click `Load`, a file selection dialog will pop up
- You can `visualize` JSON files.
- `Zoom-in/Zoom-out` is possible.
- `GPU/CPU` usage is displayed.
- You can click on the bar to see `More info`.
- You can `open/close`  the panel.

# Features to develop
- Changed to responsive scale when zooming in/out
- Fixed left bar
- Bar multi-selection (individual multi-selection, section dragging)
- Make the color a bit prettier
- Capture

# Reference
For the structure of this file, see the two Gits below.
- https://github.com/rebornix/vscode-webview-react
- https://github.com/iskenxan/sheepy-fp-guide

### CODE REVIEW

**Comment by llFreetimell** [comment]
Parent Issue : #28

**Comment by llFreetimell** [comment]
@parkjisu6239 

Maybe... https://github.com/microsoft/vscode-extension-samples/tree/main/notebook-renderer-react-sample/src can give us some clue...? 🙏

File: src/BarChart/src/features/barchart/Board.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        return Object.keys(this.state.data).map((key) => {
+            return  <Level processName={key} utility={this.state.utility[key]} data={this.state.data[key]} key={key} rulerCnt={this.state.rulerCnt} clickBar={this.clickBar}/>
+        });
+    }
+
+    render() {
+        return (
+        <div className="main-container">
+            <nav>
+                <button>Record</button>
```
Comment (llFreetimell) [inline src/BarChart/src/features/barchart/Board.js:119]: I think we do not need this button :)

File: src/BarChart/src/features/barchart/Board.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                <button>Save</button>
```
Comment (llFreetimell) [inline src/BarChart/src/features/barchart/Board.js:120]: This button is also not need I think :D

File: src/BarChart/src/features/barchart/Board.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            const data = JSON.parse(reader.result).traceEvents
+            this.processData(data)
+        }
+        reader.readAsText(file, /* optional */ "euc-kr");
+    }
+
+    processData(data) {
+        const processedData = {}
+        const backgroundColor = {}
+        const utility = {"CPU Operators": 0, "GPU Operators": 0}
```
Comment (llFreetimell) [inline src/BarChart/src/features/barchart/Board.js:67]: `CPU Operators` and `GPU operators` are not fixed name :(
It should be read from json file...

Comment (parkjisu6239) [reply src/BarChart/src/features/barchart/Board.js:67]: all right! Then I'll fix it to show % in all pids! : )

File: src/BarChart/src/features/barchart/Board.js (hunk 2 above)
Comment (parkjisu6239) [reply src/BarChart/src/features/barchart/Board.js:120]: thank you! So, instead of those two buttons, can I just have one capture button?

Comment (llFreetimell) [reply src/BarChart/src/features/barchart/Board.js:120]: That would be graet! :D

**Comment by parkjisu6239** [comment]
I have a question!
Currently, black letters are basic on a white background, so is it better to do it in dark mode?
The vscode is usually black background, so I'll modify it if you want dark mode.
Or support both light and dark modes...

**Comment by llFreetimell** [comment]
> is it better to do it in dark mode?

For me yes but not sure for others :)

> Or support both light and dark modes...

It would be really really good to have both of them!
But if this consumes much time, we can support it later because there are more important features to support :D
Anyway, I will add this to candidate feature lists in #28!

**Comment by llFreetimell** [comment]
I have a question...!
I failed to run this draft because

- `yarn` is not installed --> after installation, passed
- `cliapp` is not installed --> after installation, python2 grammar was found so failed (`print r`)
- `ttystatus` is not installed
- ...

If we install ONE-vscode, these packages will be automatically installed?

**Comment by parkjisu6239** [comment]
I think it's a problem that I haven't encountered because the current development was carried out in the window environment. 😥

@wnsdud4197 is checking the contents right now, and we will check whether it is running in the Linux environment as soon as possible! I think this is the most important thing! 🔥🔥

**Comment by CODe1995** [comment]
# How to install yarn on ubuntu18.04

>  Note: When using apt install to install yarn, "There are no scenarios" error occurs.
> Be sure to follow the instructions below.



1. `$ sudo apt-get update`

2. `$ sudo apt-get install -y nodejs`

3. Check if the program is installed.

   ```
   $ node -v
   $ npm -v
   ```

   

4. `$ npm install yarn`

5. `$ yarn -v`

**Comment by llFreetimell** [comment]
Which file should I open and press `F5`?

**Comment by CODe1995** [comment]
> Which file should I open and press F5?

`F5` is just a shortcut to debug the whole program. After pressing `F5`, when the debugging window is launched, press `ctrl`+`shift`+`p` to execute the command.

**Comment by llFreetimell** [comment]
First of all, `F5` is redirected according to the file format.
  - If I open `tfconfig.json`, vscode try to find debugging extension for json
  - If I open `extension.ts`, choose debugger type and open new window
 
And when I opened with `extension.ts`, new window is opened but our command did not appear in command lists(Ctrl+Shift+P)

**Comment by CODe1995** [comment]
@llFreetimell 

`.env` file has been added to `.gitignore`, so you won't see the run screen. We will update it right away, so please wait😀

**Comment by llFreetimell** [comment]
Can we update `react-scripts` version to 4.0.3?
Current merged code requires `eslint` 7.32.0 but `react-scripts` 2.1.3 cannot use it ... :(

**Comment by parkjisu6239** [comment]
I will close this draft because it has been discussed enough, and i have received the desired answer.
Additional content is here #93 .

**Pull request closed without merging**